### ISSUE TITLE
get_proposals: get 'inf' and 'nan'  ， why pre_nms_limit = 6000 ?

### ISSUE THREAD

**Issue by kavinbj** [Viredery/tf-eager-fasterrcnn#4]
初学者，好多不懂的，疫情期间一直在看这个代码。
我的运行环境：
macbook pro: 非GPU环境
python 3.7.3
tensorflow: 2.0.0

贴一下我的测试代码inspect_model.py：

import numpy as np
from detection.utils.misc import *
import os
from detection.datasets.utils import get_original_image
from detection.datasets import coco, data_generator
from detection.models.detectors import faster_rcnn
import visualize
import matplotlib.pyplot as plt

img_mean = (123.675, 116.28, 103.53)
img_std = (1., 1., 1.)

val_dataset = coco.CocoDataSet('./COCO2017/', 'val',
                               flip_ratio=0,
                               pad_mode='fixed',
                               mean=img_mean,
                               std=img_std,
                               scale=(800, 1024))

train_generator = data_generator.DataGenerator(val_dataset)

img, img_meta, bboxes, labels = val_dataset[0]
batch_imgs = tf.Variable(np.expand_dims(img, 0))
batch_metas = tf.Variable(np.expand_dims(img_meta, 0))

model = faster_rcnn.FasterRCNN(
    num_classes=len(val_dataset.get_categories()))

rgb_img = np.round(img + img_mean)

ori_img = get_original_image(img, img_meta, img_mean)

proposals = model.simple_test_rpn(img, img_meta)

print('proposals', proposals)

打印信息：
loading annotations into memory...
Done (t=0.52s)
creating index...
index created!
2020-03-15 15:08:44.383147: I tensorflow/core/platform/cpu_feature_guard.cc:145] This TensorFlow binary is optimized with Intel(R) MKL-DNN to use the following CPU instructions in performance critical operations:  SSE4.1 SSE4.2 AVX AVX2 FMA
To enable them in non-MKL-DNN operations, rebuild TensorFlow with the appropriate compiler flags.
2020-03-15 15:08:44.383574: I tensorflow/core/common_runtime/process_util.cc:115] Creating new thread pool with default inter op setting: 8. Tune using inter_op_parallelism_threads for best performance.
proposals tf.Tensor(
[[ 0.  0.  1. nan  1.]
 [ 0.  0.  0. nan nan]
 [ 0.  0.  1. nan  1.]
 ...
 [ 0.  0.  1. nan  1.]
 [ 0.  0.  1. nan  1.]
 [ 0.  0.  0. nan nan]], shape=(2000, 5), dtype=float32)
Segmentation fault: 11

问题：
1、为什么在get_proposals时会出现大量的'non'
       //Get refined anchors
        proposals = transforms.delta2bbox(anchors, rpn_deltas, 
                                          self.target_means, self.target_stds)
   追到
    // Apply delta
    center_y += delta[:, 0] * height
    center_x += delta[:, 1] * width
    height *= tf.exp(delta[:, 2])
    width *= tf.exp(delta[:, 3])

  这段代码产生大量的'inf' 数据，导致最后proposals出现大量nan
  rpn_deltas数据需要进行一些处理吗？直接进行 tf.exp会出错。

2、为什么要选择6000？  
      // Improve performance
        pre_nms_limit = min(6000, anchors.shape[0])
        ix = tf.nn.top_k(rpn_probs, pre_nms_limit, sorted=True).indices
     有五个特征图，分别有 (256, 256), (128, 128), (64, 64), (32, 32), (16, 16)
    总共 rpn_class_logits: shape=(n, 261888, 2)，如果是只选6000个，那基本都是在P2层 (256, 256)，所在层对应anchor的 scale size是非常小的。实际发现rpn_probs中为1的数据，有75000多个。这样导致P3-P5层的anchor没有选用上，也会导致最后计算ROI数据不准确。

**Comment by Viredery** [comment]
先说第二个问题
ix = tf.nn.top_k(rpn_probs, pre_nms_limit, sorted=True).indices
这行代码的意思是，我们先对这20w Anchors根据score进行排序，选择了score最高的6000个，将其他的过滤掉，因此并不存在你说的“基本都是在P2层 (256, 256)”。
这里的6000相当于超参数，你设置10000也行，2000也行。

第一个问题可能和“Segmentation fault: 11”有关。我刚试下了用cpu跑，我也是这个情况。
我觉得产生这个问题的原因是模型没有初始化。你既没有加载训练好的模型，也没有对卷积权重进行最简单的初始化，比如Xavier和MSRA初始化方法。

**Comment by kavinbj** [comment]
问题2：
tf.nn.top_k(input, k, name=None)
解释：这个函数的作用是返回 input 中每行最大的 k 个数，并且返回它们所在位置的索引。
经过softmax得到的rpn_probs，里面只有0或者1两种score（我打印出来的结果）， 如果进行排序，能够将P3-P5层，都排到前面6000个里面吗？这一段比较疑惑。

问题1：应该是这你说的这个问题，我没有加载训练好的模型。之前好像有点出错，因为我在研究一下

**Comment by Viredery** [comment]
“里面只有0或者1两种score（我打印出来的结果）”，这是第一个batch的前向传播得到的结果，模型会根据loss计算梯度，进行反向传播。
在整个训练的过程，会不断地正向和反向传播，去让模型学到我们希望它学到的东西。

**Comment by kavinbj** [comment]
通过加载已经训练好的模型，已经可以正确的执行测试代码。
我现在的疑问是：
假如不加载训练好的模型，如何进行初始化设置，如何从零进行数据训练？有没有相关文档可以参考一下，网上这一块说的都不太清楚。

**Comment by Viredery** [comment]
一般做目标检测，至少会加载ImageNet初始化的骨干网络权重。
完全从零开始训练的话，可以参考Rethinking ImageNet Pre-training这篇文章

代码方面的话，卷积初始化和BN的部分需要做些改动
